### PR TITLE
Implement scoring board and embed it into `game` and `game_view`

### LINKED ISSUE #542: [Game Play] [WIP] Add `scoring_board` class
Proposal: We would like to introduce a point based system. The basic idea is that you get points by doing things, e.g. eating food, collecting items and power-ups and, most importantly, killing other players. A winner may be decided according to the scores.

Scoring board class scans events happened at every tick and updates scores for the players and also records the time past since the start of the game.

### PR DESCRIPTION
**Pull request by EvoLandEco**
Fixes #542

### CODE REVIEW

**Review by TheoPannetier — CHANGES_REQUESTED** [review]
I'm a bit out of the loop with the direction gameplay is taking, but I enjoy this idea of a scoring board a lot. Looking forward to see where it goes!
I enjoy the clean structure of the class, and the scope of the tests blocks is well defined, and most importantly, it is easy to read your expectations form the test.

There are however some major issues that I would like to see addressed before merging.

1. This works largely exceeds the scope of #542, and I'd like to see more issues created, each corresponding to a single identifiable problem. For example, "there should be a scoring board class" (#542), "scoring board should have a timer", "scoring board should keep track of each player's scores", etc.
2. Some of your tests expect that erroneous values are ignored. This can easily cause silent issues, so it would be more desirable that the tests expect an error in such cases.
3. I don't like that you use a number to store the winning player, I such much room for this going terribly wrong. I suggest using an enum class instead, but I'm not sure it's the best solution, so I suggest you wait for @richelbilderbeek 's review before changing anything there.

File: game.cpp
Comment (EvoLandEco) [reply game.cpp:392]: We can omit Function Group, and make it straight forward: 
```c++
// BEGIN ** Event **
things
// END ** Event **

// BEGIN ** Scoring Board **
things
// END ** Scoring Board **
```

Does it look better?

Comment (TheoPannetier) [reply game.cpp:392]: Looks nice indeed!

File: game.cpp
Comment (EvoLandEco) [reply game.cpp:2221]: Sure, I'll create more issues later, it's just that many ideas came up when I was writing the class. Our team is trying to work in the sprint mode, so we implement things really fast and are too eager to achieve the goal ASAP, sometimes we forget to break a big issue to small ones and attempt to fix a lot of things in only one PR

File: game.cpp
Comment (EvoLandEco) [reply game.cpp:2230]: Thanks, I'll change it

File: scoring_board.cpp
Comment (EvoLandEco) [reply scoring_board.cpp:22]: I agree that doing this might plant hidden danger here. A workaround may be that we create a enum class of winners, give it a method `winner.get_index()` to convert it to index.

File: scoring_board.cpp
Comment (EvoLandEco) [reply scoring_board.cpp:35]: I'll further break this issue down, make a task list here, I feel like I need to fix many issues to get the scoring board really working

File: scoring_board.cpp
Comment (EvoLandEco) [reply scoring_board.cpp:68]: Good to know this! Indeed, it is better be informed about a malfunctioning, rather than fixing it automatically

File: scoring_board.h
Comment (EvoLandEco) [reply scoring_board.h:15]:
```c++
if (score < 0) throw "Score should be non-negative!"
```
Maybe adding this?

File: game.cpp
Comment (TheoPannetier) [reply game.cpp:2221]: Yeah, I understand the eagerness :)
Writing new issues + tests for all ideas and intermediate steps can be quite a speed bump to getting the code written indeed. But in my experience this method does speed things up in the long run, because
1. It makes it possible for other people to join and pick up the problem where you left it, thus multiplying your speed by the nb of contributors ;)
2. Debugging becomes much, much easier and faster if you work in small, self-contained issues.

Comment (TheoPannetier) [reply game.cpp:2221]: Just in case: no need to delete anything from the code! 
Just make new issues describing what should be in the class, refer to them in the tests you wrote, and maybe even close the issues if they're already solved.

File: scoring_board.cpp
Comment (TheoPannetier) [reply scoring_board.cpp:22]: I like this solution :)

File: scoring_board.h
Comment (TheoPannetier) [reply scoring_board.h:15]: Yup, that should do the trick!

Comment (TheoPannetier) [reply scoring_board.h:15]:
```
if (score < 0) throw std::logic_error("Score should be non-negative!");
```

**Comment by TheoPannetier** [comment]
@EvoLandEco why is this PR is labelled with `depends`?

We use this label to indicate that an issue/PR can't be solved currently, because it requires another issue or PR to be closed first.
For example, #463 depends on #381 because invulnerability status first needs to be implemented if we then want to make it disappear after some time, so with the depends label we let contributors now that they would need to solve the latter before tackling the former.

I feel this is unjustified here?

File: scoring_board.h
Comment (swom) [inline scoring_board.h:15]:
```suggestion
    void set_score_player1(int score) {m_score_player1 = score;};
```
The value need to be negative for us to check that it is negative, `unsigend int` unfortunately can never be negative, so we will never catch the cultpript.

File: scoring_board.h
Comment (swom) [reply scoring_board.h:45]: #550

**Review by swom — CHANGES_REQUESTED** [review]

**Comment by richelbilderbeek** [comment]
@EvoLandEco : I asked Theo and Oliver to take over the `tiangjian` branch, as this PR slows us the team down for already a month, and I cannot be sure to see you soon. 

I know it may feel a bit rude to have your branch taken over, but I hope you will agree that in this context it's better for the team. We just want to get that scoreboard in :-)

File: scoring_board.cpp
Comment (TheoPannetier) [reply scoring_board.cpp]: Made a commit to fix this using throw :)

File: scoring_board.cpp
Comment (ollyturner) [inline scoring_board.cpp:43]: What we could do is make the winner an enum class that can take the value 0, 1, 2 or none.

Comment (TheoPannetier) [reply scoring_board.cpp:43]: #590

File: scoring_board.cpp
Comment (TheoPannetier) [inline scoring_board.cpp:61]: I disagree with the implementation here; I think the game should throw an error if someone tries to set score to a negative value.
```suggestion
        sb.set_score_player1(-500);
        assert(sb.get_score_player1() == 500);
        sb.set_score_player2(-500);
        assert(sb.get_score_player2() == 500);
        sb.set_score_player3(-500);
        assert(sb.get_score_player3() == 500);
```

Comment (TheoPannetier) [reply scoring_board.cpp:61]: #591

File: scoring_board.cpp
Comment (TheoPannetier) [reply scoring_board.cpp:71]: #591

File: scoring_board.cpp
Comment (TheoPannetier) [reply scoring_board.cpp:87]: #590

**Review by TheoPannetier — APPROVED** [review]

**Pull request merged**
